[CodeCompletion] Update member completion to handle ambiguous and invalid base expressions

nathawes opened the pull request:
This hooks up member completion to the new `typeCheckForCodeCompletion` API to generate completions from all solutions the constraint solver produces – including ones requiring fixes – rather than relying on a single solution being applied to the AST (if any). This lets us still produce member completion results when the base expression is ambiguous or invalid.

Whenever typeCheckExpression is called on an expression containing a code completion expression and a CompletionCallback has been set, each solution formed is passed to the callback so the type and decl of the base expression can be extracted along with any other info needed to generate member completions.

Commit:
[CodeCompletion] Pass the CodeCompletionExpr rather than just the base expression to the DotExpr completion callback.

Commit:
[CodeCompletion] Add initial implementation of solver-based member completion.

This hooks up member completion to the new typeCheckForCodeCompletion API to
generate completions from all solutions the constraint solver produces (include
ones requiring fixes) rather than relying purely the single solution being
applied to the AST (if any). This lets us still give completion results in
ambiguous and invalid code.

xedin committed:
[CodeCompletion] Adjust `DotExprLookup::sawSolution` to use new `Solution` APIs

xedin committed:
[CSSolver] add debug logging to `solveForCodeCompletion`

xedin committed:
[CSSolver] Update expression after constraint generation while solving for code completion

Commit:
[CodeCompletion] Move CompletionCollector to ASTContext + bug fixes.

xedin committed:
[CSSolver/CodeCompletion] If pre-check fails extract code completion expression and use it

If there was an invalid reference which was caught by pre-check,
let's remove all context besides code completion itself and use
it to produce code completion results.

xedin committed:
[CSSolver] Solution: Resolve holes as underlying generic parameters for code completion

To aid code completion, we need to attempt to convert type holes
back into underlying generic parameters if possible, since type
of the code completion expression is used as "expected" (or contextual)
type so it's helpful to know what requirements it has to filter
the list of possible member candidates e.g.

```swift
func test<T: P>(_: [T]) {}

test(42.#^MEMBERS^#)
```

It's impossible to resolve `T` in this case but code completion
expression should still have a type of `[T]` instead of `[<<hole>>]`
because it helps to produce correct contextual member list based on
a conformance requirement associated with generic parameter `T`.

Commit:
[CodeCompletion] Fallback to typechecking just the completion expression in cases where typeCheckExpression is never called.

This happens when, e.g. an expression being switched on is invalid so
expression patterns in the switch cases (which may contain the completion
expression) are not checked.

Also setup the Lookup object to handle member completion in ObjC selector
expressions correctly, and fix passing the wrong expression when computing
isStaticallyDerivedMetatype().

xedin committed:
[CSSolver] Augment `solveForCodeCompletion` to indicate when it fails

xedin committed:
[TypeChecker] Mode code completion fallback to `typeCheckForCodeCompletion`

If an expression fails pre-check or constraint generation,
code completion should be performed directly on the
`CodeCompletionExpr` as a fallback. Let's factor that logic
out from `solveForCodeCompletion` and put it directly into
`typeCheckForCodeCompletion` because it's easier to establish
fault conditions there.

xedin committed:
[TypeChecker/CodeCompletion] Allow `{typeCheck, solve}ForCodeCompletion` to operate on target instead of expression

Using `SolutionApplicationTarget` make it easier to propage
contextual information and avoid mistakes of using incorrect
accessors for constraint generation.

xedin committed:
[TypeChecker/CodeCompletion] Make viability checking responsibility of `typeCheckForCodeCompletion`

There is no way to separate viability/applicability checking from
`TypeChecker::typeCheckForCodeCompletion` because multi-statement
closures could be either type-checked together with enclosing
context (e.g. when closure represents a function builder body), or
separately - when it's just a regular closure.

Due to this "duality" we need to attempt to run code completion
code to determine whether body of the closure participated in
type-check and if it didn't, fallback to a `typeCheckExpression`.

xedin committed:
[ConstraintSystem] Add `Solution::hasType(ASTNode)`

xedin committed:
[TypeChecker/CodeCompletion] Implement a fallback from code completion to `typeCheckExpression`

This fallback to `typeCheckExpression` is triggered when it's
determined that code completion expression is located inside
of a multi-statement closure and its body is going going to
participate in type-check.

xedin committed:
[TypeChecker] Safe all contextual information about completion expression

Having all information possible makes it much easier to determine
when it's appropriate to fallback to type-check code completion
expression and whether it's possible to add some more context to it.

Commit:
[CodeCompletion] Fix incorrect type relations in complete_enum_elements test.

We were reporting methods that return function types that return void (rather
than returning void directly) as being invalid in contexts that expect non-void
expressions and testing for that incorrect behavior.

Commit:
[CodeCompletion] Fix bug in logic determining whether void members are valid.

Commit:
[Parse] Perform the single expression function body transform for delayed parsing as well

We were previously only doing it when parsing up front.

Commit:
[test] Update code completion test with improved type relation result.

Commit:
[CodeCompletion][test] Update test/IDE/complete_exception.swift with improved type relation.

Commit:
[Sema] TypeCheckNodeAtLoc: strip out the implicit return the parser adds to single expression bodies for void-returning functions

This makes TypeCheckNodeAtLoc match the regular function body typechecking
behavior and fixes a crash in test/IDE/complete_value_expr.swift

Commit:
[test] Add member completion tests where the base has errors, or is ambiguous.

xedin committed:
[TypeChecker/CodeCompletion] Move hole -> archetype conversion logic into `sawSolution`

Instead of exposing archetypes during solution generation, let's
do that based on solution data when a type of code completion
expression is requested by code completion callback.

xedin committed:
[ConstraintSystem] Allow code completion type to be a hole

Code completion expression type could be a hole if and only
if there is obosolutely no contextual information available
e.g. `let _ = .#^MEMBER^#`

xedin committed:
[CodeCompletion] Don't attempt lookup if base type couldn't be determined

If base type couldn't be determined (e.g. because base expression
is an invalid reference), let's not attempt to do a lookup since
it wouldn't produce any useful results anyway.

Commit:
[Test][CodeCompletion] Consistently check for the 'unknown' type relation when there's no expected type.

The tests were matching the previous implementation's output, which sometimes produced 'unknown' and sometimes 'unrelated' in cases where there was no expected type from the context.

Commit:
[CodeCompletion][NFC] Add doc comments and rename symbols for clarity in the new member completion implementation.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform

nathawes commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -55,6 +55,11 @@ struct TypeWitnessAndDecl;
+
+namespace constraints {
+  class Solution;
+}
```
Oops, the changes in this file aren't needed anymore.

xedin commented:
/cc @hborla @DougGregor

xedin commented:
@swift-ci please test macOS platform

Commit:
Remove unused function declaration and the forward declarations of types it used.

Commit:
[CodeCompletion] Handle member completions within the OriginalExprs of ErrorExprs if present.

We weren't walking into ErrorExprs in typeCheckForCodeCompletion previously,
so never found the CodeCompletionExpr.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform

rintaro commented:
Toolchain https://ci.swift.org/job/swift-PR-toolchain-osx/677/artifact/branch-master/swift-PR-33749-677-osx.tar.gz

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -5795,9 +5811,203 @@ static void addConditionalCompilationFlags(ASTContext &Ctx,
+static void deliverCompletionResults(CodeCompletionContext &CompletionContext,
+                                     CompletionLookup &Lookup,
+                                     SourceFile &SF,
+                                     CodeCompletionConsumer &Consumer) {
```
I believe you just factored the code out from `CodeCompletionCallbacksImpl`. So i'm not reviewing the actual logic here. If you made any behavior changes, let me know.

nathawes replied:
I think I may have changed everything to use the passed-in SF, while previously it was using P.SF in some places and CurDeclContext->getParentSourceFile() in others. Is that safe?

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -5795,9 +5811,203 @@ static void addConditionalCompilationFlags(ASTContext &Ctx,
+    DotExprLookup Lookup(DotLoc, CurDeclContext, CodeCompleteTokenExpr);
+    llvm::SaveAndRestore<CompletionCollector*>
+      CompletionCollector(Context.CompletionCallback, &Lookup);
+    typeCheckContextAt(CurDeclContext, CompletionLoc);
```
(Not in this PR) When we migrate everything, we should rename this to something else :)
e.g.
```cpp
performCompletionTypeChecking(CurDeclContext, ComletionLoc, Lookup);
```

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -5795,9 +5811,203 @@ static void addConditionalCompilationFlags(ASTContext &Ctx,
+
+    // This (hopefully) only happens in cases where the expression isn't
+    // typechecked during normal compilation either (e.g. member completion in a
+    // switch case where there switched value is invalid). Having normal
```
According the the language reference, it's called "control expression"
https://docs.swift.org/swift-book/ReferenceManual/Statements.html#ID436

rintaro commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -592,20 +592,282 @@ TypeChecker::getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
+  ContextFinder contextAnalyzer(expr);
+  expr->walk(contextAnalyzer);
+
+  // If there was no completion expr (e.g. if the code completion location is
+  // within an ErrorExpr without an valid subexpr due to parser error recovery)
+  // bail.
+  if (!contextAnalyzer.hasCompletionExpr())
+    return false;
```
The original expression in `ErrorExpr` is now walked into. If you know other cases, please update the comment.

nathawes replied:
Right, I think this branch is impossible to get into at the moment since we're only doing this for member completion. Do you know if once we change the other completion kinds to use this approach there will ever be a case where no CodeCompletionExpr is formed within an ErrorExpr? I assumed this was possible, e.g. if the completion location is in the middle of some nonsensical tokens between statement starting tokens, but if we're always guaranteed to have a CodeCompletionExpr regardless of completion kind I think this should be an assertion.

rintaro replied:
I don’t know off the top of my head. But I think we should keep this `return false` path, just to be on the safe side.

rintaro commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -592,20 +592,282 @@ TypeChecker::getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
+  // and regular closures which are type-checked separately.
+
+  {
+    // First, pre-check the expression, validating any types that occur in the
+    // expression and folding sequence expressions.
+    auto failedPreCheck = ConstraintSystem::preCheckExpression(
+        expr, DC,
+        /*replaceInvalidRefsWithErrors=*/true);
+
+    target.setExpr(expr);
+
+    if (failedPreCheck)
+      return false;
```
If `preCheckExpression()`, it falls back to normal type checking which calls `preCheckExpression()` again. @xedin Is it safe to call `preCheckExpression()` on already pre-checked expressions?

xedin replied:
Yeah, it's effectively a noop.

rintaro commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -592,20 +592,282 @@ TypeChecker::getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
+                                               /*contextualType=*/Type(),
+                                               /*isDiscarded=*/true);
+
+    (void)solveForCodeCompletion(completionTarget);
```
Assert the result is not `CompletionResult::NotApplicable`?

rintaro reviewed (commented):
Looks great in general 👍

nathawes commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -703,3 +925,114 @@ LookupResult
+    ReferencedDecl = SelectedOverload->choice.getDeclOrNull();
+
+  auto Key = std::make_pair(BaseTy, ReferencedDecl);
+  auto Ret = BaseToSolutionIdx.insert({Key, Solutions.size()});
```
It's the value here, rather than the key. If the key is not already in the BaseToSolutionIdx map we will insert a new entry into the Solutions vector, and Solutions.size() will be its index.

rintaro replied:
Ah, sorry. I misread the code 😓

nathawes commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -44,6 +44,64 @@ namespace swift {
+    SourceLoc DotLoc;
+    DeclContext *DC;
+    CodeCompletionExpr *CompletionExpr;
+    SmallVector<SolutionInfo, 4> Solutions;
```
Yeah, I'm not super happy with most of the names. I was hoping to come up with something better for "CompletionCollector" and "DotExprLookup" too.

rintaro replied:
How about
`CompletionCollector` -> `TypeCheckCompletionCallback`
`DotExprLookup` -> `DotExprTypeCheckCompletionCallback`
`Solution` -> `Result`

nathawes commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -44,6 +44,64 @@ namespace swift {
+    /// Called for each solution produced while  type-checking an expression containing a code
+    /// completion expression.
+    void sawSolution(const constraints::Solution &solution) override;
+  };
```
Makes sense to me, particularly as we'll hopefully be adding a bunch more of these.

nathawes commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -1526,6 +1530,8 @@ void CodeCompletionContext::sortCompletionResults(
+
 class CodeCompletionCallbacksImpl : public CodeCompletionCallbacks {
   CodeCompletionContext &CompletionContext;
   std::vector<RequestedCachedModule> RequestedModules;
```
Oh maybe not - thanks!

rintaro commented on test/IDE/complete_enum_elements.swift:
```diff
@@ -133,7 +133,7 @@ enum FooEnum: CaseIterable {
 // FOO_ENUM_DOT_INVALID-NEXT: Decl[EnumElement]/CurrNominal: Foo2[#FooEnum#]{{; name=.+$}}
 // FOO_ENUM_DOT_INVALID-NEXT: Decl[StaticVar]/CurrNominal: alias1[#FooEnum#]{{; name=.+$}}
-// FOO_ENUM_DOT_INVALID-NEXT: Decl[InstanceMethod]/CurrNominal/TypeRelation[Invalid]: hash({#(self): FooEnum#})[#(into: inout Hasher) -> Void#]{{; name=.+$}}
+// FOO_ENUM_DOT_INVALID-NEXT: Decl[InstanceMethod]/CurrNominal: hash({#(self): FooEnum#})[#(into: inout Hasher) -> Void#]{{; name=.+$}}
```
This makes `FOO_ENUM_DOT_INVALID` the same as `FOO_ENUM_DOT`. Just remove `FOO_ENUM_DOT_INVALID` and modify the `FileCheck` argument to use `FOO_ENUM_DOT`

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -5795,9 +5811,203 @@ static void addConditionalCompilationFlags(ASTContext &Ctx,
+                                   DCForModules);
+}
+
+void DotExprLookup::performLookup(ide::CodeCompletionContext &CompletionCtx,
+                                  CodeCompletionConsumer &Consumer,
+                                  bool IsInSelector) const {
```
It's not clear to me why the logic in this function should be in `DotExprLookup` especially considering this is declared in `Sema`. How about `DotExprLookup` just have an accessor for `Solutions` and inline this logic in `trySolverCompletion`?

Commit:
[CodeCompletion] Remove old DotExpr completion handling code - it's unreachable now.

Commit:
[CodeCompletion] Rename a few types/methods for clarity plus other small refactorings (NFC)

Also put subclasses of TypeCheckCompletionCallback into their own header.

Commit:
[CodeCompletion] Remove unused forward declaration and fix up an out-of-date comment (NFC)

nathawes commented:
@swift-ci clean test

Commit:
[CodeCompletion] Remove unused field in CodeCompletionCallbacksImpl (NFC)

nathawes commented:
@swift-ci please test

Commit:
[CodeCompletion] Deduplicate the two isMemberCompletion functions in ParseExpr.cpp and ParseDecl.cpp

Also:
- propagate the Solution -> Result rename to Solution parameter of deliverDotExprResults
- fixup header comment in CodeCompletionTypeChecking.h

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test Linux

rintaro reviewed (approved):
LGTM
